spec/spec.md

```diff
@@ -89,6 +89,23 @@ is reported if:
 The type of the `Return` expression is `void`. It returns control from the
 current procedure to its caller.
 
+#### `Unreachable`
+
+```grammar
+expr += (Unreachable)
+```
+
+Marks a control-flow path as unreachable. In case program execution does reach
+the `Unreachable` expression, the program immediately terminates. A compiler
+*may* report an error if it can statically detect that control-flow can reach
+an `Unreachable` expression within a procedure.
+
+> TODO: consider lifting the requirement that the program must terminate, in
+> order to allow for the compiler optimizing the affected control-flow paths
+> away
```

Comment:
Right now it states that `program execution` terminates, I'd imagine in the case of wanting to not terminate for the purposes of optimization, that would be `program analysis` of some variety that's perhaps doing some sort of pseudo evaluation?

Reply:
> that would be program analysis of some variety that's perhaps doing some sort of pseudo evaluation?

Hm, I'm not sure I get what you mean.

In my mind, the statement "the program immediately terminates" requires that execution of program terminates -- *always*. This precludes optimizations where the compiler produces no code for control-flow paths leading up to `Unreachable`, under the assumption that `Unreachable` really does mean "never reached".

Having said that, I'm no longer sure that such an optimization is really are a good idea. Perhaps in an explicit `-d:danger`-like mode, where the is allowed to violate the specification, but not otherwise

I'll remove the TODO.

Reply:
Ah, yeah, your interpretation makes sense.